.github/workflows/trivy-security-scan.yml

```diff
@@ -1,15 +1,40 @@
 ##
-# This action runs trivy container and repository vulnerability
-# scanner for docker images and cargo packages.
+# This action runs Trivy container and repository vulnerability
+# scanner for Docker images and filesystem.
 ##
 
 name: trivy-security-scan
 
 on:
   repository_dispatch:
-    types: [ trivy-scan-dispatch ]
+    types: [trivy-scan-dispatch]
+  schedule:
+    - cron: '0 0 * * *'  # Runs daily at midnight UTC
 
 jobs:
+  update-trivy-db:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Setup oras
+        uses: oras-project/setup-oras@9c92598691bfef1424de2f8fae81941568f5889c # v1.21
+
+      - name: Get current date
+        id: date
+        run: echo "date=$(date +'%Y-%m-%d')" >> $GITHUB_OUTPUT
+
+      - name: Download and extract the vulnerability DB
+        run: |
+          mkdir -p $GITHUB_WORKSPACE/.cache/trivy/db
+          oras pull ghcr.io/aquasecurity/trivy-db:2
+          tar -xzf db.tar.gz -C $GITHUB_WORKSPACE/.cache/trivy/db
+          rm db.tar.gz
+
+      - name: Cache DBs
+        uses: actions/cache@v4
+        with:
+          path: ${{ github.workspace }}/.cache/trivy
+          key: cache-trivy-${{ steps.date.outputs.date }}
```

Comment:
1. This creates a race condition between the update job and the first workflow that runs and tries to use the new date’s cache.

Some of our workflows run on the merge queue, which can delay them by a few hours. We also have staff that are working around midnight UTC, due to time zones.

2. GitHub also says cron jobs can be skipped entirely under load:
> The schedule event can be delayed during periods of high loads of GitHub Actions workflow runs. High load times include the start of every hour. If the load is sufficiently high enough, some queued jobs may be dropped. To decrease the chance of delay, schedule your workflow to run at a different time of the hour.

Could we just have a single cache instead?

That would also avoid filling up our cache quota with unused copies of similar images. If we make this change, we could also remove the date step.

```suggestion
          key: cache-trivy
```

Another alternative would be keeping the date, but adding `restore-keys: cache-trivy` wherever we access the cache. This would fall back to an older cache if today’s isn’t available.

Reply:
@teor2345 why? It's just taking the current date and will cache that, if the date changes it downloads a new DB the next day.

Reply:
> @teor2345 why? It's just taking the current date and will cache that, if the date changes it downloads a new DB the next day. If we have a single cache how will it know to update the DB?

My original comment was based on having a cron job, which would eventually update the cache, even if the key didn’t contain the date. But that’s changed now, so it doesn’t matter any more.

> With regards to caching quota, GitHub's caching mechanism automatically manages cache retention, typically removing caches that haven't been accessed within the last week, or when the total cache size exceeds the 10GB limit. 

Yes, I know 🙂

> The Trivy DB is ~10-30MB in [size](https://aquasecurity.github.io/trivy/v0.38/docs/vulnerability/examples/db/).

Aha, I didn’t know it was that small. Daily updates to separate caches aren’t too bad then.

> For the race condition, i have adjusted the jobs with dependencies. Albeit, the second job does not depend on the update db job, and the `update-trivy-db` would complete first since the image build takes much longer.

I think how it works now is better.